Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstructionModel.py

```diff
@@ -0,0 +1,165 @@
+# limitations under the License.
+
+from clr import AddReference
+AddReference("System")
+AddReference("QuantConnect.Algorithm")
+AddReference("QuantConnect.Common")
+AddReference("QuantConnect.Indicators")
+
+from System import *
+from QuantConnect import *
+from QuantConnect.Indicators import *
+from QuantConnect.Algorithm import *
+from QuantConnect.Algorithm.Framework import *
+from QuantConnect.Algorithm.Framework.Alphas import *
+from QuantConnect.Algorithm.Framework.Portfolio import PortfolioTarget
+from datetime import timedelta
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize
+
+### <summary>
+### Provides an implementation of Mean-Variance portfolio optimization based on modern portfolio theory.
+### The interval of weights in optimization method can be changed based on the long-short algorithm.
+### The default model uses the last three months daily price to calculate the optimal weight 
+### with the weight range from -1 to 1 and minimize the portfolio variance with a target return of 2%
+### </summary>
+class MeanVarianceOptimizationPortfolioConstructionModel:
+    def __init__(self, *args, **kwargs):
+        """ 
+        Initialize the model
+        Args:
+            resolution: The resolution of the history price 
+            period(int): The time interval of history price to calculate the weight 
+            minimum_weight(float): The lower bounds on portfolio weights
+            maximum_weight(float): The upper bounds on portfolio weights
+            target_return(float): The target portfolio return
+            optimization_method(method): Method used to compute the portfolio weights
+        """
+        self.resolution = kwargs['resolution'] if 'resolution' in kwargs else Resolution.Daily
+        self.period = kwargs['period'] if 'period' in kwargs else 63
+        self.minimum_weight = kwargs['minimum_weight'] if 'minimum_weight' in kwargs else -1
+        self.maximum_weight = kwargs['maximum_weight'] if 'maximum_weight' in kwargs else 1
+        self.target_return = kwargs['target_return'] if 'target_return' is kwargs else 2
+        self.optimization_method = kwargs['optimization_method'] if 'optimization_method' in kwargs else self.minimum_variance
+        self.symbolDataBySymbol = {}
+
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm: The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        symbols = [insight.Symbol for insight in insights]
+        if len(symbols) == 0:
+            return []
+
+        for insight in insights:
+            symbolData = self.symbolDataBySymbol.get(insight.Symbol)
+            symbolData.Add(algorithm.Time, insight.Magnitude)
```

Comment:
Looks like this is building history of _expected_ returns for the covariance calculations -- I think we should be using historical returns.

Also might want to do a null check here on `insight.Magnitude` since it's not a required field, and throw if it is null w/ a good error message stating that the alpha model must emit a non-null `insight.Magnitude` to be used with this MVO PC model.